### DISCUSSION TITLE
Running LISEM on Linux

### DISCUSSION THREAD

**Discussion by geografin** [bastianvandenbout/LISEM#12] [Ideas]
Hi, LISEM community members! 
I would like to use and run LISEM on linux without any interface (it should be possible, as PCRaster is nicely working on Linux).
I am rather experienced PCRaster user, so the specific language logic is clear.
Does it exist a common way to use it on linux (I don't see it in downloads, so probably not) and what would you advice me to begin with?
Should I compile the model from scratch and then develop utilities to manage .map format data or I should take windows version and use it without windows interface?

**Comment by bastianvandenbout** [comment]
Hi Geografin,

In principle it is possible to compile on linux (I managed before), but it is quite involved and not well supported at the moment.
Independent of OS, there are quite some external dependencies. On Windows, MSYS2 Helps a lot with this ( a linux-style package manager for c++ libraries on windows), and we have a pre-build set of dll and include files available for compilation.
For linux, you have to prepare all the dependencies yourself. Not all of them are available through package managers, and it can be quite cumbersome.
The OpenMVG and OpenMVS libraries are a bit complicated, so you might simple disable any code that uses them.
On the other hand, OTB and ITK are used, and OTB uses ITK 4, while ITK has moved on to version 5. You can either use ITK 4 or alter OTB yourself to work with ITK 5 (there is already a branch on their repo for this).
We dont actually use pcraster for our algorithms and scripting anymore. We still support pcraster file format, but algorithms and scripting are different.

Good luck, and please post any issues you encounter!

**Comment by smoors** [comment]
hi, i'm a co-maintainer of [EasyBuild](https://docs.easybuild.io/), and we're also very interested in getting LISEM running on Linux, specifically on HPC clusters. getting LISEM into EasyBuild would make your software available to a wide range of users and researchers.

i've managed to get all the dependencies installed.
LISEM itself gets past the CMake configure step, and it builds until ~70% (GCC compiler toolchain).
however, there seems to be quite some Windows-specific code that would need to be ported, and i don't have much experience with this.

how much work would it be to port this to Linux?
is there any chance for you to implement Linux support in the near future?

**Reply by bastianvandenbout** [reply]
Hi Smoors, 
Hmm, yes its not so easy for the new version of the software. Classic compiles rather easily with the dependencies, but for the LISEM 0.2.4 I have not yet had the time to invest in making it linux-friendly. It is possible, but both due to poor support from my our side and the xomplexity of letting all the dependencies play nicely together, its not so easy now.
I am happy to help with particulars, but wont be able to provide a working solution in the near future,

Kind regards,
Bastian

**Comment by feefladder** [comment]
Just some notes while trying to install on a relatively fresh xubuntu 23.10:

- minimum CMake in `third_party/g3m/CMakeLists.txt` is `2.8`, which gives a deprecation warning
- install required packages (MSYS2 is windows-specific) [qt](https://stackoverflow.com/a/67415291/14681457) [qt multimedia](https://stackoverflow.com/a/71641234/14681457) [itk](https://docs.itk.org/en/latest/download.html#package-managers): 
  ```
  sudo apt install libboost-all-dev libeigen3-dev libgmp-dev qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools libgdal-dev libinsighttoolkit5-dev
  ```
actually libinsighttoolkit5-dev already installs a lot of qt5 stuff, so maybe qt5 is redundant.
OTB need to get file, because it is only in debian, also CMAKE says it should be version 7.0.0, get it from [the index](https://www.orfeo-toolbox.org/packages/archives/OTB/) Why is 7.0.0 > 800 MB? Does it also work with 9.0.0?

gives this error:
```
-- Checking for C++14 compiler
-- Checking for C++14 compiler - available
Co
ler info: 
C -Compiler: /usr/bin/c++
C++ -Compiler: /usr/bin/cc
Detected LISEM version: 0.2.3
CMake Deprecation Warning at third_party/g3m/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- GMP libs: /usr/lib/x86_64-linux-gnu/libgmp.so /usr/lib/x86_64-linux-gnu/libgmpxx.so
Did not Find External Dir
CMake Error at cmakeFindModules/externals.cmake:36 (message):
  Specify the folder containing external libraries using EXTERNAL_DIR
Call Stack (most recent call first):
  CMakeLists.txt:127 (include)


-- Configuring incomplete, errors occurred!
```


line `include(externals)` links to file `cmakeFindModules/externals.cmake`

instead of normal `cmake ..` use `cmake -DLISEM_USE_EXTERNALDIR=false ..`

However, [OTB 7.0.0](https://gitlab.orfeo-toolbox.org/orfeotoolbox/otb/-/tags?page=2) needs to be manually compiled. For this:

```
mkdir -p local/otb-7.0.0
cd local/otb-7.0.0
wget https://gitlab.orfeo-toolbox.org/orfeotoolbox/otb/-/archive/7.0.0/otb-7.0.0.tar.gz
tar -xvf otb=7.0.0.tar.gz
mkdir build
cd build
# cmake .. # error about ITKNeuralNetworks
# Github: https://github.com/InsightSoftwareConsortium/ITKNeuralNetworks
# says don't use for anything serious => uncomment line in local/Modules/ThirdParty/ITK/otb-module-init.cmake
sed -i 's/(ITKNeuralNetworks)/# $1/' ../Modules/ThirdParty/ITK/otb-module-init.cmake

# why not use a later version?

# need openthreads
sudo apt install libopenthreads-dev

# now, need to install ossim...
# Git: https://github.com/ossimlabs/ossim
# 🤣🤣 We hope to soon have a Windows build.bat file and world peace. 🤣🤣
# go back to local
cd ../../
wget https://github.com/ossimlabs/ossim/releases/download/v2.12.0/ossim-2.12.0.tar.gz
tar -xvf ossim-2.12.tar.gz
mv ossim-2.12 ossim #see for why: https://github.com/ossimlabs/ossim/issues/296

sudo apt install libgeos++-dev libtiff-dev libjpeg-dev

cd ossim
scripts/build.sh
cd ..

# at this point I got multiple defined errors and gave up

```
